### DISCUSSION TITLE
tutorials/physics/kinematic_character_2d

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#146] [User-contributed notes]
# tutorials/physics/kinematic_character_2d

Introduction: Yes, the name sounds strange. "Kinematic Character". What is that? The reason for the name is that, when physics engines came out, they were called "Dynamics" engines (because they de...

https://docs.godotengine.org/en/stable/tutorials/physics/kinematic_character_2d.html

**Comment by ReinhardVA** [comment]
Why didn't you use move_and_collide when implement walking? Because i'm trying to implement it and i couldn't do it.
Either the character is moving slowly or gravity is not working but there is always be a problem.

**Reply by LoganWyborn** [reply]
If your character is using MoveAndCollide, you have to add additional logic to implement the "sliding" behaviour. The character is colliding with the floor due to the downward velocity and thus is stopping. instead of:
```
MoveAndCollide(velocity * (float)delta);
```
you will use something more like:
```
var collision = MoveAndCollide(velocity * (float)delta);
if (collision is not null)
{
	velocity = velocity.Slide(collision.GetNormal());
	MoveAndCollide(velocity * (float)delta);
}
```
collision here is a KinematicCollision2D that contains info about what MoveAndCollide hit. The body is given it's initial velocity, then attempts to move. It collides with the floor, then we "slide" along the floor based on the result of our collision.

**Comment by suumpmolk** [comment]
The examples here mixes `move_and_collide()` and `move_and_slide()` which is confusing.

In the last example, if `move_and_slide()` takes delta into account, should it be 

```
velocity.y += GRAVITY
```

and not 

```
velocity.y += delta * GRAVITY
```

?

**Reply by timothyqiu** [reply]
This is not true. `move_and_collide()` and `move_and_slide()` are not necessarily related to `delta`, only unit matters.

`move_and_collide()` needs a movement vector, in pixels. `move_and_slide()` needs a speed vector, in pixels per second.

Whether `GRAVITY` needs to be multiplied by `delta` is determined by the unit of `GRAVITY`. It is not specified in the example code. But "gravity" by definition is the acceleration, it's the amount of change in velocity, in pixels per second per second. So multiplying it by seconds is the expected pixels per second.

**Reply by suumpmolk** [reply]
That makes sense, thanks for the explanation!

**Comment by IsaacFehr** [comment]
I'm quite confused by this page.  Is it up to date?  I looked at the [2d/platformer](https://github.com/godotengine/godot-demo-projects/tree/master/2d/platformer) demo, which this page says has an example of a dynamic character controller, but that example looks like it uses a kinematic controller with `move_and_slide()`.

Is the dynamic character controller example now in [2d/physics_platformer](https://github.com/godotengine/godot-demo-projects/tree/master/2d/physics_platformer)

Or am I misunderstanding?  My takeaway from this page was that a dynamic controller uses a `RigidBody2d` with `_integrate_forces()` instead of `CharacterBody2D` and `move_and_slide()` or `move_and_collide()`, which is for kinematic characters.

**Reply by Calinou** [reply]
KinematicBody2D was more or less the Godot 3.x name of CharacterBody2D.

I think `2d/platformer` used RigidBody2D for its character a long time ago, but it was switched to KinematicBody2D throughout the 3.x cycle and has remained so on 4.x.
